File: go/core/interfaces/lifecycle.go
package interfaces

import "context"

// Lifecycle is implemented by any client/component with coordinated startup and
// shutdown, so a lifecycle.Manager can Start them in registration order and Stop
// them in reverse for graceful shutdown. It mirrors the mothership client-lifecycle
// contract.
type Lifecycle interface {
	// Start initializes the component and establishes its connection/resources. It
	// should block until the component is ready or ctx is cancelled.
	Start(ctx context.Context) error

	// Stop gracefully releases the component's resources and completes in-flight
	// work.
	Stop(ctx context.Context) error
}

// HealthChecker reports a component's health for Kubernetes probes: Liveness
// answers "is it alive?" (restart the pod if not) and Readiness "is it ready to
// serve?" (remove the pod from the load balancer if not).
type HealthChecker interface {
	// Liveness reports whether the component is alive (a lightweight self-check that
	// does not depend on external systems).
	Liveness(ctx context.Context) error

	// Readiness reports whether the component can serve traffic — typically by
	// verifying connectivity (e.g. a Ping).
	Readiness(ctx context.Context) error
}

// Client is the base contract for a resilient, lifecycle-managed client (gRPC,
// database pool, Connect server): it can be started/stopped by a lifecycle.Manager
// and health-probed by Kubernetes. It composes Lifecycle + HealthChecker per the
// mothership client platform.
type Client interface {
	// Lifecycle provides coordinated Start/Stop.
	Lifecycle
	// HealthChecker provides the Kubernetes Liveness/Readiness probes.
	HealthChecker
}

File: python/techai_webutils/src/techai_webutils/core/interfaces/lifecycle.py
"""Client lifecycle + health contracts (the Python analog of Go's lifecycle stack).

These mirror Go's ``interfaces.Lifecycle`` / ``HealthChecker`` / ``Client``.
Python's lifecycle idiom is the async context manager managed by
``contextlib.AsyncExitStack`` at a service's composition root: it enters
resources in registration order and exits them in reverse — exactly the role Go's
``foundation/lifecycle.Manager`` plays. These Protocols give that idiom a *named*
contract and add the Kubernetes health-probe surface (liveness/readiness) the
composition root and health endpoints check.

Because these are ``@runtime_checkable`` Protocols, any backend that already
implements ``__aenter__``/``__aexit__`` (the S3, SQS, lock, and DLQ clients do)
structurally satisfies ``ManagedResource`` without inheriting from it.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Protocol, Self, runtime_checkable

if TYPE_CHECKING:
    from types import TracebackType


@runtime_checkable
class ManagedResource(Protocol):
    """A client/component whose lifecycle is an async context manager.

    Connection-owning backends acquire their resource in ``__aenter__`` and release it
    in ``__aexit__``, so a composition root manages them uniformly with an
    ``AsyncExitStack`` (enter in order, exit in reverse). It is the Python analog of
    Go's ``interfaces.Lifecycle`` (Start/Stop); ``AsyncExitStack`` is the analog of
    ``foundation/lifecycle.Manager``.
    """

    async def __aenter__(self) -> Self:
        """Enter the async context, acquiring the resource; return self."""
        ...

    async def __aexit__(
        self,
        exc_type: type[BaseException] | None,
        exc_val: BaseException | None,
        exc_tb: TracebackType | None,
    ) -> bool | None:
        """Exit the async context, releasing any owned resource."""
        ...


@runtime_checkable
class HealthChecker(Protocol):
    """Reports a component's health for Kubernetes probes.

    ``liveness`` answers "is it alive?" — a lightweight self-check that does not depend
    on external systems (restart the pod if it raises). ``readiness`` answers "is it
    ready to serve?" — typically verifies connectivity, e.g. a ping (deroute the pod if
    it raises). Both raise on failure and return ``None`` on success. Mirrors Go's
    ``interfaces.HealthChecker``.
    """

    async def liveness(self) -> None:
        """Raise if the component is not alive; return None if it is."""
        ...

    async def readiness(self) -> None:
        """Raise if the component cannot serve traffic; return None if it can."""
        ...


@runtime_checkable
class Client(ManagedResource, HealthChecker, Protocol):
    """The base contract for a resilient, lifecycle-managed client.

    It owns an async resource lifecycle (``ManagedResource``) and answers Kubernetes
    health probes (``HealthChecker``). Mirrors Go's ``interfaces.Client``.
    """
